juju/juju
# Implement "juju deploy --attach-storage"

## Description of change

This branch adds the --attach-storage flag to the "juju deploy" command, along with the corresponding API and API server changes. The state changes were made in a previous branch. With this change, it is possible to attach existing storage to the initial unit of an application.

There are a few fixes bundled in here as separate commits:
 - revert the changes to handling Dead volume/filesystem attachments. Machine storage attachments go from Alive to Dying, and Dying to removed. There is no Dead state.
 - modify the preconditions for block device updates so that they can still occur for Dying, but not Dead, machines. A Dying machine may still require block device updates to clean up storage (e.g. to detach a volume-backed filesystem).
 - add short-circuit removal of volume-backed filesystems. Once a volume-backed filesystem is detached, it has no responsible storage provisioner. Thus, destroying a detached volume-backed filesystem should cause it to be removed immediately; the contents will be destroyed along with the volume.

## QA steps

1. juju bootstrap (tested with aws and google; google assumed below)
2. juju deploy postgresql --storage pgdata=gce,10G
3. juju ssh postgresql/0
$ sudo -u postgres psql
> CREATE TABLE FOO(bar VARCHAR);
> INSERT INTO FOO VALUES('baz');
4. juju remove-application postgresql
5. juju deploy postgresql pg2 --attach-storage pgdata/0 --to zone=<same availability zone as postgresql/0 was deployed to>
6. juju ssh pg2/0
$ sudo -u postgres psql
> SELECT * FROM FOO;
(check for previously inserted row)
7. juju destroy-controller -y

NOTE: it is a current limitation that the AZ must be explicitly specified. This limitation will be addressed before 2.3 is released.

## Documentation changes

Yes, we will need to add documentation of the new --attach-storage flag. It accepts a list of detached storage IDs, which will be attached to the initial unit of the application. This flag may not be used with "-n".

## Bug reference

None.
